[PAC][CodeGen][ELF][AArch64] Support signed GOT

kovdan01 opened the pull request:
This re-applies #96164 after revert in #102434.

Support the following relocations and assembly operators:

- `R_AARCH64_AUTH_ADR_GOT_PAGE` (`:got_auth:` for `adrp`)
- `R_AARCH64_AUTH_LD64_GOT_LO12_NC` (`:got_auth_lo12:` for `ldr`)
- `R_AARCH64_AUTH_GOT_ADD_LO12_NC` (`:got_auth_lo12:` for `add`)

`LOADgotAUTH` pseudo-instruction is introduced which is later expanded to actual instruction sequence like the following.

```
adrp x16, :got_auth:sym
add x16, x16, :got_auth_lo12:sym
ldr x0, [x16]
autia x0, x16
```

If a resign is requested, like below, `LOADgotPAC` pseudo is used, and GOT load is lowered similarly to `LOADgotAUTH`.

```
@var = global i32 0
define ptr @resign_globalvar() {
  ret ptr ptrauth (ptr @var, i32 3, i64 43)
}
```

If FPAC bit is not set and resign is requested, a check+trap sequence similar to one used for `AUT` pseudo is emitted.

Both SelectionDAG and GlobalISel are suppported.
For FastISel, we fall back to SelectionDAG.

Tests starting with 'ptrauth-' have corresponding variants w/o this prefix.

See also specification
https://github.com/ARM-software/abi-aa/blob/main/pauthabielf64/pauthabielf64.rst#appendix-signed-got

kovdan01 commented:
I was unable to reproduce the original buildbot failure https://lab.llvm.org/buildbot/#/builders/153/builds/5329. In that build, `ExecutionEngine/JITLink/AArch64/ELF_relocations.s` test fails (see https://github.com/llvm/llvm-project/pull/96164#issuecomment-2275170432).

- When triggering rebuild for that particular build, tests pass: see https://lab.llvm.org/buildbot/#/builders/153/builds/6640
- When triggering force build for branch with changes re-applied, tests pass: see https://lab.llvm.org/buildbot/#/builders/153/builds/6580
- Locally, I've cross-compiled llvm for aarch64 and ran the failing test under qemu for hundreds of times (just in case it's some unstable test which fails not at every time) - the test passes every time

Maybe the failure was somehow related to #100854 (which touched the failing test) being merged several commits earlier than #96164. The build, corresponding to that PR, also passed - see https://lab.llvm.org/buildbot/#/builders/153/builds/5325.

Unfortunately, I can't now say what was the reason of the original failure. Even though the src dir and the build dir were not cleaned before build, that should not make any difference in this case.

Would be glad to see everyone's thoughts on probable reasons of original build failure if you have some. As for now - I just re-submit the patch (currently treating the failure as a strange instability which was not reproduced neither on buildbot nor locally) with a couple of changes described below.

kovdan01 commented:
Comparing to original PR #96164, this one has a couple of differences:

- `LOADgotAUTH` expansion code is moved from expand pseudos pass to asm printer. For other similar PAC-related pseudos, like `LOADgotPAC`, the expansion code is already placed in asm printer, so it's probably better to keep things consistent and add new code to the same place (given that I don't see strong reasons to put it earlier in pipeline in expand pseudos pass).

- If resign is requested, we cannot just do `aut{i|d}a` and then `pac{i|d}a` w/o FPAC bit set. In such a case, if authentication failure occurs, we'll just silently sign the pointer, while we want to check if authentication was successful and somehow fail if it was not. With FPAC, this is done implicitly, w/o that - we need to do that manually. The check+trap sequence code is similar to one used in `AUTPAC` pseudo expansion.

kovdan01 commented:
> Comparing to original PR #96164, this one has a couple of differences:

One new difference proposed with latest PR update - do not rely on bitmask pauthabi version value to determine if codegen needs to use signed GOT. It's specific to llvm_linux platform. Instead, check if module flag "ptrauth-elf-got" is present and equal to 1.

kovdan01 commented:
@MaskRay @tmatheson-arm @smithp35 Would be glad to see your feedback on this

kovdan01 commented:
The latest PR update fixes the following issue. With signed ELF GOT enabled, the linker looks at the symbol type to choose between keys IA (for STT_FUNC) and DA (for other types). Symbols for functions not defined in the module have STT_NOTYPE type by default. This makes linker to emit signing schema with DA key (instead of IA) for corresponding R_AARCH64_AUTH_GLOB_DAT dynamic reloc. To avoid that, force all function symbols used in the module to have STT_FUNC type. See https://github.com/ARM-software/abi-aa/blob/main/pauthabielf64/pauthabielf64.rst#default-signing-schema

@MaskRay could you please clarify if this is OK from linker's perspective? Does lld rely on the fact that undefined functions are STT_NOTYPE or is this OK to have them STT_FUNC? I've ran test-suite built with pauth enabled (including signed GOT), and nothing seems to break, but maybe there are some non-obvious issues which might appear under some conditions.

Would be glad to see everyone's feedback on this @tmatheson-arm @smithp35

asl commented on llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp:
```diff
@@ -923,6 +928,22 @@ void AArch64AsmPrinter::emitEndOfAsmFile(Module &M) {
+        M.getModuleFlag("ptrauth-elf-got"));
+    if (PtrAuthELFGOTFlag && PtrAuthELFGOTFlag->getZExtValue() == 1)
+      for (const GlobalValue &GV : M.global_values())
+        if (!GV.use_empty() && GV.getValueType()->isFunctionTy() &&
```
Can it be simplified down to `isa<Function>(GV)`? Also, what happens with aliases?

kovdan01 replied:
> Can it be simplified down to `isa<Function>(GV)`?

Yes, it can, thanks for suggestion! Fixed in cc9681063f645a808fec804253c7f0e598377b98

> Also, what happens with aliases?

Aliases must point to a definition (if I don't miss smth). Symbols for functions with definitions will anyway be marked as `STT_FUNC`. The point of the loop is to mark all function symbols which only have declarations as `STT_FUNC`. Yes, the loop also covers defined functions (not function aliases), but there is no issue in "double-marking" them - it'll not result in duplicate annotation in assembly. A test for alias to a defined function is present in llvm/test/CodeGen/AArch64/ptrauth-got-abuse.ll.

smithp35 commented:
> The latest PR update fixes the following issue. With signed ELF GOT enabled, the linker looks at the symbol type to choose between keys IA (for STT_FUNC) and DA (for other types). Symbols for functions not defined in the module have STT_NOTYPE type by default. This makes linker to emit signing schema with DA key (instead of IA) for corresponding R_AARCH64_AUTH_GLOB_DAT dynamic reloc. To avoid that, force all function symbols used in the module to have STT_FUNC type. See https://github.com/ARM-software/abi-aa/blob/main/pauthabielf64/pauthabielf64.rst#default-signing-schema
> 
> @MaskRay could you please clarify if this is OK from linker's perspective? Does lld rely on the fact that undefined functions are STT_NOTYPE or is this OK to have them STT_FUNC? I've ran test-suite built with pauth enabled (including signed GOT), and nothing seems to break, but maybe there are some non-obvious issues which might appear under some conditions.
> 
> Would be glad to see everyone's feedback on this @tmatheson-arm @smithp35

Closest I can find in terms of use of STT_NOTYPE is this thread in the GABI https://groups.google.com/g/generic-abi/c/_ZPPq_c8FSQ which to my reading suggests that emitting the type on the reference could be a good thing as it would permit static linker's to diagnose mismatches in symbol types such as STT_FUNC and STT_OBJECT. However I think the discussion stopped short of requiring it. It doesn't however forbid it.

There's a bug report in GCC advocating it, with a patch that doesn't seem to have been applied https://gcc.gnu.org/bugzilla/show_bug.cgi?id=35514.

From what I can see, it looks like the main advantage and disadvantage of setting the type on the symbol reference is that the static linker has to do some extra work to make sure that a combination of two references to sym one with STT_NOTYPE and one with any other type preserves the type. It also has to decide what to do if there is a mismatch of types on the reference and definition. It could ignore the reference, or could warn.

The most likely alternative is to provide alternative relocation types for code and non-code references. This is essentially moving the symbol type into the relocation code though.

kovdan01 commented:
> > The latest PR update fixes the following issue. With signed ELF GOT enabled, the linker looks at the symbol type to choose between keys IA (for STT_FUNC) and DA (for other types). Symbols for functions not defined in the module have STT_NOTYPE type by default. This makes linker to emit signing schema with DA key (instead of IA) for corresponding R_AARCH64_AUTH_GLOB_DAT dynamic reloc. To avoid that, force all function symbols used in the module to have STT_FUNC type. See https://github.com/ARM-software/abi-aa/blob/main/pauthabielf64/pauthabielf64.rst#default-signing-schema
> > @MaskRay could you please clarify if this is OK from linker's perspective? Does lld rely on the fact that undefined functions are STT_NOTYPE or is this OK to have them STT_FUNC? I've ran test-suite built with pauth enabled (including signed GOT), and nothing seems to break, but maybe there are some non-obvious issues which might appear under some conditions.
> > Would be glad to see everyone's feedback on this @tmatheson-arm @smithp35
> 
> Closest I can find in terms of use of STT_NOTYPE is this thread in the GABI https://groups.google.com/g/generic-abi/c/_ZPPq_c8FSQ which to my reading suggests that emitting the type on the reference could be a good thing as it would permit static linker's to diagnose mismatches in symbol types such as STT_FUNC and STT_OBJECT. However I think the discussion stopped short of requiring it. It doesn't however forbid it.
> 
> There's a bug report in GCC advocating it, with a patch that doesn't seem to have been applied https://gcc.gnu.org/bugzilla/show_bug.cgi?id=35514.
> 
> From what I can see, it looks like the main advantage and disadvantage of setting the type on the symbol reference is that the static linker has to do some extra work to make sure that a combination of two references to sym one with STT_NOTYPE and one with any other type preserves the type. It also has to decide what to do if there is a mismatch of types on the reference and definition. It could ignore the reference, or could warn.
> 
> The most likely alternative is to provide alternative relocation types for code and non-code references. This is essentially moving the symbol type into the relocation code though.

@smithp35 Thanks for your feedback! Yes, providing alternative relocation types is a viable solution for this. But, as far as I see, the current way it's resolved now (by unconditionally setting `STT_FUNC` type even for undefined functions) is also not prohibited.

@MaskRay Could you please give your feedback on the problem? Would be glad to hear which solution looks better from linker's perspective.

kovdan01 committed:
[PAC][CodeGen][ELF][AArch64] Support signed GOT

This re-applies #96164 after revert in #102434.

Support the following relocations and assembly operators:

- `R_AARCH64_AUTH_ADR_GOT_PAGE` (`:got_auth:` for `adrp`)
- `R_AARCH64_AUTH_LD64_GOT_LO12_NC` (`:got_auth_lo12:` for `ldr`)
- `R_AARCH64_AUTH_GOT_ADD_LO12_NC` (`:got_auth_lo12:` for `add`)

`LOADgotAUTH` pseudo-instruction is introduced which is later expanded to
actual instruction sequence like the following.

```
adrp x16, :got_auth:sym
add x16, x16, :got_auth_lo12:sym
ldr x0, [x16]
autia x0, x16
```

If a resign is requested, like below, `LOADgotPAC` pseudo is used, and GOT
load is lowered similarly to `LOADgotAUTH`.

```
@var = global i32 0
define ptr @resign_globalvar() {
  ret ptr ptrauth (ptr @var, i32 3, i64 43)
}
```

If FPAC bit is not set and resign is requested, a check+trap sequence
similar to one used for `AUT` pseudo is emitted.

Both SelectionDAG and GlobalISel are suppported.
For FastISel, we fall back to SelectionDAG.

Tests starting with 'ptrauth-' have corresponding variants w/o this prefix.

See also specification
https://github.com/ARM-software/abi-aa/blob/main/pauthabielf64/pauthabielf64.rst#appendix-signed-got

kovdan01 committed:
[PAC][CodeGen] Use `emitPtrauthCheckAuthenticatedValue` in `LowerMOVaddrPAC`

kovdan01 committed:
[PAC][CodeGen] Fix handling of extern_weak symbols with signed GOT

Before executing auth instruction on a signed address from a GOT slot,
we must ensure that the address is not zero since zero addresses are not
signed to preserve zero-checks working as expected. A GOT slot might be
filled with zero when we have an undefined weak symbol.

kovdan01 committed:
[PAC][CodeGen] Emit trap sequence w/o FPAC on LOADgotAUTH expansion

kovdan01 commented:
Comments regarding latest update:

1. For emitting check sequence after auth instruction when no FPAC, use `emitPtrauthCheckAuthenticatedValue` introduced in #110702. See 118e2fb4dd1504fda4732bf069dbfcdb0edff5aa

2. When handling extern weak symbols during `LOADgotAUTH` expansion, check the pointer against null before auth instruction (null pointer is not signed and should not be authenticated). See 9a5f1ef5a92507bc6760f81f93faa5938f78eba1

3. In `LOADgotAUTH` expansion, emit check sequence just like it is done before resign in `LOADgotPAC` expansion. Though `LOADgotAUTH` does not contain resign, we do not know how the authenticated pointer would be used in code later, so we should conservatively emit check sequence in `LOADgotAUTH` right after auth instruction. See 61bae48816e81d9f8adc4f5af0e290c047466c96

kovdan01 commented:
@MaskRay @tmatheson-arm @smithp35 Would be glad to see your feedback on this.

@MaskRay Earlier, there was a question regarding setting `STT_FUNC` type for even undefined function symbols, see [issuecomment-2360450900](https://github.com/llvm/llvm-project/pull/105798/#issuecomment-2360450900). There was some subsequent discussion, see [issuecomment-2371987887](https://github.com/llvm/llvm-project/pull/105798/#issuecomment-2371987887). Could you please give a feedback on this issue from linker's perspective?

kovdan01 commented:
I've re-opened this as #113811 since I had to change the source branch to allow stacked PR (codegen support for signed GOT with tiny code model #113812 and for signed TLSDESC #113813 depend on this). The new PR is identical to this one, and the commit history is preserved.

Would be glad to continue discussion in #113811 - closing this one in favor of that.
